crawl-ref/source/movement.cc

```diff
@@ -454,6 +502,251 @@ bool prompt_dangerous_portal(dungeon_feature_type ftype)
+    beam.is_targeting    = true;
+    beam.fire();
+
+    const monster* valid_target = nullptr;
+
+    // Iterate the tracer to see if the first visible target is a hostile mons.
+    for (coord_def p : beam.path_taken)
+    {
+        // Don't lunge if our tracer path is broken by deep water, lava,
+        // teleport traps, etc., before it reaches a monster.
+        if (!feat_is_traversable(grd(p))
+            && !(grd(p) == DNGN_SHALLOW_WATER))
+        {
+            break;
+        }
+        // Don't lunge if the tracer path is broken by something solid or
+        // transparent: doors, grates, etc.
+        if (cell_is_solid(p) || you.trans_wall_blocking(p))
+            break;
+
+        const monster* mon = monster_at(p);
+        if (!mon)
+            continue;
+        // Don't lunge at invis mons, but allow the tracer to keep going.
+        else if (mon && !you.can_see(*mon))
+            continue;
+        // Don't lunge if the closest mons is non-hostile or a plant.
+        else if (mon && (mon->friendly()
+                         || mon->neutral()
+                         || mons_is_firewood(*mon)))
+        {
+            break;
+        }
+        // Okay, the first mons along the tracer is a valid target.
+        else if (mon)
+        {
+            valid_target = mon;
+            break;
+        }
+    }
+    if (!valid_target)
+        return spret::fail;
+
+    // Reset the beam target to the actual lunge_target distance.
+    beam.target = lunge_target;
+
+    // Don't lunge if the player's tile is being targeted, somehow.
+    if (beam.target == you.pos())
+        return spret::fail;
+
+    // Don't lunge if it would take us away from a beholder.
+    const monster* beholder = you.get_beholder(beam.target);
+    if (beholder)
+    {
+        clear_messages();
+        mprf("You cannot lunge away from %s!",
+             beholder->name(DESC_THE, true).c_str());
+        return spret::fail;
+    }
+
+    // Don't lunge if it would take us toward a fearmonger.
+    const monster* fearmonger = you.get_fearmonger(beam.target);
+    if (fearmonger)
+    {
+        clear_messages();
+        mprf("You cannot lunge closer to %s!",
+             fearmonger->name(DESC_THE, true).c_str());
+        return spret::fail;
+    }
+
+    // Don't lunge if it would land us on top of a monster.
+    const monster* mons = monster_at(beam.target);
+    if (mons)
+    {
+        if (!you.can_see(*mons))
+        {
+            // .. if it was in the way and invisible, notify the player.
+            clear_messages();
+            mpr("Something unexpectedly blocked you, preventing you from lunging!");
+        }
+        return spret::fail;
+    }
+
+    // Don't lunge if the target tile has a dangerous (!FFT_SOLID) feature:
+    if (feat_is_lava(grd(beam.target)))
+        return spret::fail;
+    else if (grd(beam.target) == DNGN_DEEP_WATER
+             || grd(beam.target) == DNGN_TOXIC_BOG)
```

Comment:
At the time I wrote this, I had a concern that lunging over deep water / lava could potentially lead to unavoidable deaths due to flight expiry, as any valid lunge over these features would lead to the player being two tiles away from safe terrain.

4e7fcc has since removed the old instadeath codepath here, and this concern is now obviated, so, okay, sounds good.

Reply:
Welcome to the future, we've got fun and games